File: impl/go/transforms.go
package twill

import "strings"

// Transform utilities (SPEC §10.8, "Transforms").

// Transform is the composed transform value.
const Transform = "var(--tw-rotate-x,) var(--tw-rotate-y,) var(--tw-rotate-z,) var(--tw-skew-x,) var(--tw-skew-y,)"

const translate2D = "var(--tw-translate-x) var(--tw-translate-y)"
const translate3D = translate2D + " var(--tw-translate-z)"
const scale2D = "var(--tw-scale-x) var(--tw-scale-y)"
const scale3D = scale2D + " var(--tw-scale-z)"

func transformRegistrations() []Node {
	return []Node{
		Property("--tw-rotate-x", nil), Property("--tw-rotate-y", nil), Property("--tw-rotate-z", nil),
		Property("--tw-skew-x", nil), Property("--tw-skew-y", nil),
	}
}

func translateRegistrations() []Node {
	zero := strPtr("0")
	return []Node{Property("--tw-translate-x", zero), Property("--tw-translate-y", zero), Property("--tw-translate-z", zero)}
}

func scaleRegistrations() []Node {
	one := strPtr("1")
	return []Node{Property("--tw-scale-x", one), Property("--tw-scale-y", one), Property("--tw-scale-z", one)}
}

var transformOrigins = [][2]string{
	{"center", "center"}, {"top", "top"}, {"top-right", "top right"}, {"right", "right"},
	{"bottom-right", "bottom right"}, {"bottom", "bottom"}, {"bottom-left", "bottom left"},
	{"left", "left"}, {"top-left", "top left"},
}

// RegisterTransformUtilities registers the transform utilities.
func RegisterTransformUtilities(u *Utilities, theme *Theme) {
	stat := func(name string, declarations ...[2]string) { StaticUtility(u, name, Declarations(declarations)) }
	statFn := func(name string, fn func() []Node) { StaticUtilityFn(u, name, fn) }
	fn := func(root string, desc FunctionalUtilityDescription) { FunctionalUtility(u, theme, root, desc) }
	degrees := func(v *CandidateValue) (string, bool) { return v.Value + "deg", IsPositiveInteger(v.Value) }
	percentage := func(v *CandidateValue) (string, bool) { return v.Value + "%", IsPositiveInteger(v.Value) }
	single := func(property string) func(string, *string) []Node {
		return func(value string, _ *string) []Node { return []Node{Decl(property, value)} }
	}

	stat("transform-none", [2]string{"transform", "none"})
	statFn("transform-gpu", func() []Node {
		return append(transformRegistrations(), Decl("transform", "translateZ(0) "+Transform))
	})
	statFn("transform-cpu", func() []Node { return append(transformRegistrations(), Decl("transform", Transform)) })
	fn("transform", FunctionalUtilityDescription{Handle: single("transform")})
	stat("transform-flat", [2]string{"transform-style", "flat"})
	stat("transform-3d", [2]string{"transform-style", "preserve-3d"})
	stat("backface-visible", [2]string{"backface-visibility", "visible"})
	stat("backface-hidden", [2]string{"backface-visibility", "hidden"})

	stat("perspective-none", [2]string{"perspective", "none"})
	fn("perspective", FunctionalUtilityDescription{ThemeKeys: []string{"--perspective"}, Handle: single("perspective")})
	for _, o := range transformOrigins {
		stat("perspective-origin-"+o[0], [2]string{"perspective-origin", o[1]})
		stat("origin-"+o[0], [2]string{"transform-origin", o[1]})
	}
	fn("perspective-origin", FunctionalUtilityDescription{Handle: single("perspective-origin")})
	fn("origin", FunctionalUtilityDescription{Handle: single("transform-origin")})

	translate := func(name string, variables []string, composed string, fractions bool) {
		handle := func(value string) []Node {
			nodes := translateRegistrations()
			for _, v := range variables {
				nodes = append(nodes, Decl(v, value))
			}
			return append(nodes, Decl("translate", composed))
		}
		SpacingUtility(u, theme, name, []string{"--translate", "--spacing"}, handle, true, fractions)
		if fractions {
			statFn(name+"-full", func() []Node { return handle("100%") })
			statFn("-"+name+"-full", func() []Node { return handle("-100%") })
		}
	}
	translate("translate", []string{"--tw-translate-x", "--tw-translate-y"}, translate2D, true)
	translate("translate-x", []string{"--tw-translate-x"}, translate2D, true)
	translate("translate-y", []string{"--tw-translate-y"}, translate2D, true)
	translate("translate-z", []string{"--tw-translate-z"}, translate3D, false)
	statFn("translate-3d", func() []Node { return append(translateRegistrations(), Decl("translate", translate3D)) })
	stat("translate-none", [2]string{"translate", "none"})

	scaleCompile := func(c *Candidate, negative bool) ([]Node, CompileStatus) {
		if c.Value == nil || c.Modifier != nil {
			return nil, NotHandled
		}
		if c.Value.Kind == ValueArbitrary {
			if negative {
				return nil, NotHandled
			}
			return []Node{Decl("scale", c.Value.Value)}, Handled
		}
		value, ok := theme.Resolve(strPtr(c.Value.Value), []string{"--scale"}, 0)
		if !ok {
			if value, ok = percentage(c.Value); !ok {
				return nil, NotHandled
			}
		}
		if negative {
			value = "calc(" + value + " * -1)"
		}
		return append(scaleRegistrations(),
			Decl("--tw-scale-x", value), Decl("--tw-scale-y", value), Decl("--tw-scale-z", value),
			Decl("scale", scale2D)), Handled
	}
	u.Functional("scale", func(c *Candidate) ([]Node, CompileStatus) { return scaleCompile(c, false) }, nil)
	u.Functional("-scale", func(c *Candidate) ([]Node, CompileStatus) { return scaleCompile(c, true) }, nil)
	for _, axis := range []string{"x", "y", "z"} {
		axis := axis
		composed := scale2D
		if axis == "z" {
			composed = scale3D
		}
		fn("scale-"+axis, FunctionalUtilityDescription{
			ThemeKeys: []string{"--scale"}, SupportsNegative: true, HandleBareValue: percentage,
			Handle: func(value string, _ *string) []Node {
				return append(scaleRegistrations(), Decl("--tw-scale-"+axis, value), Decl("scale", composed))
			},
		})
	}
	statFn("scale-3d", func() []Node { return append(scaleRegistrations(), Decl("scale", scale3D)) })
	stat("scale-none", [2]string{"scale", "none"})

	fn("rotate", FunctionalUtilityDescription{ThemeKeys: []string{"--rotate"}, SupportsNegative: true, HandleBareValue: degrees, Handle: single("rotate")})
	stat("rotate-none", [2]string{"rotate", "none"})
	for _, axis := range []string{"x", "y", "z"} {
		axis := axis
		fn("rotate-"+axis, FunctionalUtilityDescription{
			ThemeKeys: []string{"--rotate"}, SupportsNegative: true, HandleBareValue: degrees,
			Handle: func(value string, _ *string) []Node {
				return append(transformRegistrations(),
					Decl("--tw-rotate-"+axis, "rotate"+strings.ToUpper(axis)+"("+value+")"),
					Decl("transform", Transform))
			},
		})
	}

	skew := func(name string, axes []string) {
		fn(name, FunctionalUtilityDescription{
			ThemeKeys: []string{"--skew"}, SupportsNegative: true, HandleBareValue: degrees,
			Handle: func(value string, _ *string) []Node {
				nodes := transformRegistrations()
				for _, axis := range axes {
					nodes = append(nodes, Decl("--tw-skew-"+axis, "skew"+strings.ToUpper(axis)+"("+value+")"))
				}
				return append(nodes, Decl("transform", Transform))
			},
		})
	}
	skew("skew", []string{"x", "y"})
	skew("skew-x", []string{"x"})
	skew("skew-y", []string{"y"})
}

File: impl/ocaml/src/transforms.ml
(* Transform utilities (SPEC §10.8, "Transforms"). *)

open Ast
open Candidate
open Utilities

let property = Builtin_variants.property

(* The composed transform value. *)
let transform = "var(--tw-rotate-x,) var(--tw-rotate-y,) var(--tw-rotate-z,) var(--tw-skew-x,) var(--tw-skew-y,)"
let translate_2d = "var(--tw-translate-x) var(--tw-translate-y)"
let translate_3d = translate_2d ^ " var(--tw-translate-z)"
let scale_2d = "var(--tw-scale-x) var(--tw-scale-y)"
let scale_3d = scale_2d ^ " var(--tw-scale-z)"

let transform_registrations () =
  List.map (fun n -> property n None) [ "--tw-rotate-x"; "--tw-rotate-y"; "--tw-rotate-z"; "--tw-skew-x"; "--tw-skew-y" ]

let translate_registrations () =
  List.map (fun n -> property n (Some "0")) [ "--tw-translate-x"; "--tw-translate-y"; "--tw-translate-z" ]

let scale_registrations () = List.map (fun n -> property n (Some "1")) [ "--tw-scale-x"; "--tw-scale-y"; "--tw-scale-z" ]

let transform_origins =
  [ ("center", "center"); ("top", "top"); ("top-right", "top right"); ("right", "right");
    ("bottom-right", "bottom right"); ("bottom", "bottom"); ("bottom-left", "bottom left"); ("left", "left");
    ("top-left", "top left") ]

(* Registers the transform utilities. *)
let register u theme =
  let stat name declarations = static_utility u name declarations in
  let stat_fn name f = static_utility_fn u name f in
  let fn root desc = functional_utility u theme root desc in
  let degrees = bare_suffix "deg" and percentage = bare_suffix "%" in
  let single property = fun value _ -> [ decl property value ] in
  stat "transform-none" [ ("transform", "none") ];
  stat_fn "transform-gpu" (fun () -> transform_registrations () @ [ decl "transform" ("translateZ(0) " ^ transform) ]);
  stat_fn "transform-cpu" (fun () -> transform_registrations () @ [ decl "transform" transform ]);
  fn "transform" (describe (single "transform"));
  stat "transform-flat" [ ("transform-style", "flat") ];
  stat "transform-3d" [ ("transform-style", "preserve-3d") ];
  stat "backface-visible" [ ("backface-visibility", "visible") ];
  stat "backface-hidden" [ ("backface-visibility", "hidden") ];
  stat "perspective-none" [ ("perspective", "none") ];
  fn "perspective" (describe ~theme_keys:[ "--perspective" ] (single "perspective"));
  List.iter
    (fun (name, value) ->
      stat ("perspective-origin-" ^ name) [ ("perspective-origin", value) ];
      stat ("origin-" ^ name) [ ("transform-origin", value) ])
    transform_origins;
  fn "perspective-origin" (describe (single "perspective-origin"));
  fn "origin" (describe (single "transform-origin"));
  let translate name variables composed fractions =
    let handle value =
      translate_registrations () @ List.map (fun v -> decl v value) variables @ [ decl "translate" composed ]
    in
    spacing_utility u theme name [ "--translate"; "--spacing" ] handle true fractions;
    if fractions then begin
      stat_fn (name ^ "-full") (fun () -> handle "100%");
      stat_fn ("-" ^ name ^ "-full") (fun () -> handle "-100%")
    end
  in
  translate "translate" [ "--tw-translate-x"; "--tw-translate-y" ] translate_2d true;
  translate "translate-x" [ "--tw-translate-x" ] translate_2d true;
  translate "translate-y" [ "--tw-translate-y" ] translate_2d true;
  translate "translate-z" [ "--tw-translate-z" ] translate_3d false;
  stat_fn "translate-3d" (fun () -> translate_registrations () @ [ decl "translate" translate_3d ]);
  stat "translate-none" [ ("translate", "none") ];
  let scale_compile (c : candidate) negative =
    match c.cvalue with
    | None -> not_handled
    | Some _ when c.cmodifier <> None -> not_handled
    | Some { vkind = Arbitrary; value; _ } -> if negative then not_handled else handled [ decl "scale" value ]
    | Some named -> (
        let value = match Theme.resolve theme (Some named.value) [ "--scale" ] 0 with Some v -> Some v | None -> percentage named in
        match value with
        | None -> not_handled
        | Some value ->
            let value = if negative then "calc(" ^ value ^ " * -1)" else value in
            handled
              (scale_registrations ()
              @ [ decl "--tw-scale-x" value; decl "--tw-scale-y" value; decl "--tw-scale-z" value; decl "scale" scale_2d ]))
  in
  functional u "scale" (fun c -> scale_compile c false);
  functional u "-scale" (fun c -> scale_compile c true);
  List.iter
    (fun axis ->
      let composed = if axis = "z" then scale_3d else scale_2d in
      fn ("scale-" ^ axis)
        (describe ~theme_keys:[ "--scale" ] ~supports_negative:true ~handle_bare_value:percentage (fun value _ ->
             scale_registrations () @ [ decl ("--tw-scale-" ^ axis) value; decl "scale" composed ])))
    [ "x"; "y"; "z" ];
  stat_fn "scale-3d" (fun () -> scale_registrations () @ [ decl "scale" scale_3d ]);
  stat "scale-none" [ ("scale", "none") ];
  fn "rotate" (describe ~theme_keys:[ "--rotate" ] ~supports_negative:true ~handle_bare_value:degrees (single "rotate"));
  stat "rotate-none" [ ("rotate", "none") ];
  List.iter
    (fun axis ->
      fn ("rotate-" ^ axis)
        (describe ~theme_keys:[ "--rotate" ] ~supports_negative:true ~handle_bare_value:degrees (fun value _ ->
             transform_registrations ()
             @ [ decl ("--tw-rotate-" ^ axis) ("rotate" ^ String.uppercase_ascii axis ^ "(" ^ value ^ ")");
                 decl "transform" transform ])))
    [ "x"; "y"; "z" ];
  let skew name axes =
    fn name
      (describe ~theme_keys:[ "--skew" ] ~supports_negative:true ~handle_bare_value:degrees (fun value _ ->
           transform_registrations ()
           @ List.map (fun axis -> decl ("--tw-skew-" ^ axis) ("skew" ^ String.uppercase_ascii axis ^ "(" ^ value ^ ")")) axes
           @ [ decl "transform" transform ]))
  in
  skew "skew" [ "x"; "y" ];
  skew "skew-x" [ "x" ];
  skew "skew-y" [ "y" ]
